src/number_theory/legendre_symbol/quadratic_char.lean

```diff
@@ -0,0 +1,339 @@
+    { have h := finite_field.neg_one_ne_one_of_char_ne_two hF, -- `-1 ≠ 1`
+      simp only [hb', h, mul_neg, mul_one, if_false, ite_mul, neg_mul],
+      cases finite_field.pow_dichotomy hF ha with ha' ha';
+        simp only [ha', h, neg_neg, eq_self_iff_true, if_true, if_false], }, },
+end
+
+/-- The quadratic character is a homomorphism of monoids with zero. -/
+def quadratic_char_hom : F →*₀ ℤ :=
+{ to_fun := quadratic_char F, map_zero' := quadratic_char_zero, map_one' := quadratic_char_one,
+  map_mul' := quadratic_char_mul }
+
+/-- The square of the quadratic character on nonzero arguments is `1`. -/
+lemma quadratic_char_sq_one {a : F} (ha : a ≠ 0) : (quadratic_char F a)^2 = 1 :=
+by rwa [pow_two, ← quadratic_char_mul, ← pow_two, quadratic_char_sq_one']
+
+/-- The quadratic character is `1` or `-1` on nonzero arguments. -/
+lemma quadratic_char_dichotomy {a : F} (ha : a ≠ 0) :
+  quadratic_char F a = 1 ∨ quadratic_char F a = -1 :=
+(sq_eq_one_iff (quadratic_char F a)).mp (quadratic_char_sq_one ha)
+
+/-- If `F` has odd characteristic, then `quadratic_char F` takes the value `-1`. -/
+lemma quadratic_char_exists_neg_one (hF : ring_char F ≠ 2) : ∃ a, quadratic_char F a = -1 :=
+begin
+  cases (finite_field.exists_nonsquare hF) with b h₁,
+  have hb : b ≠ 0 := by { intro hf, rw hf at h₁, exact h₁ (is_square_zero F), },
+  use b,
+  simp only [quadratic_char, hb, if_false, ite_eq_right_iff],
+  tauto,
+end
+
+open_locale big_operators
+
+/-- The sum over the values of the quadratic character is zero when the characteristic is odd. -/
+lemma quadratic_char_sum_zero (hF : ring_char F ≠ 2) : ∑ (a : F), quadratic_char F a = 0 :=
+begin
+  cases (quadratic_char_exists_neg_one hF) with b hb,
+  have h₀ : b ≠ 0 := by
+  { intro hf,
+    rw [hf, quadratic_char_zero, zero_eq_neg] at hb,
+    exact one_ne_zero hb, },
+  let mul_b : F → F := λ x, b * x,
+  have h₁ : ∑ (a : F), quadratic_char F (b * a) = ∑ (a : F), quadratic_char F a :=
+  by refine fintype.sum_bijective _ (mul_left_bijective₀ b h₀) _ _ (λ x, rfl),
+  simp only [quadratic_char_mul] at h₁,
+  rw [← finset.mul_sum, hb, neg_mul, one_mul] at h₁,
+  exact eq_zero_of_neg_eq h₁,
+end
+
+end quadratic_char
+
+/-!
+### The quadratic chacter mod p
+
+We define the quadratic character on `zmod p` as a special case of `quadratic_char`.
+
+We also define the primitive quadratic characters `χ₄`on `zmod 4`
+and `χ₈`, `χ₈'` on `zmod 8`.
+-/
+
+section quad_char_mod_p
+
+/-- Define the quadratic character mod `p`, for a prime `p`. -/
+def zmod.quadratic_char (p : ℕ) [fact p.prime] : (zmod p) → ℤ := quadratic_char (zmod p)
+
+/-- The quadratic character mod `p` is a homomorphism of monoids with zero. -/
+def zmod.quadratic_char_hom (p : ℕ) [fact p.prime] : (zmod p) →*₀ ℤ :=
+@quadratic_char_hom (zmod p) _ _ _
+
+/-- Define the nontrivial quadratic character on `zmod 4`, `χ₄`.
+It corresponds to the extension `ℚ(√-1)/ℚ`. -/
+def χ₄ : (zmod 4) →*₀ ℤ :=
```

Comment:
I think this name is too generic, especially since it is in the root name space.
How about `zmod.χ₄`?

Also, I think `@[simps]` will generate a useful simp-lemma for this definition.

Reply:
That's what I had in mind, but it looks like I then forgot to use `namespace zmod`.